### PR TITLE
chore: 06-solomachine adding CheckForMisbehaviour and UpdateStateOnMisbehaviour

### PR DESCRIPTION
**Pull request by damiannolan**
## Description

- Adding `CheckForMisbehaviour` to 06-solomachine
- Adding `UpdateStateOnMisbehaviour` to 06-solomachine

ref: #878 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: modules/light-clients/06-solomachine/types/update.go [hunk 1]
```diff
@@ -84,6 +89,31 @@ func (cs ClientState) VerifyClientMessage(cdc codec.BinaryCodec, clientMsg expor
...
+	_ sdk.Context, _ codec.BinaryCodec, _ sdk.KVStore,
+	msg exported.Header, // TODO: Update to exported.ClientMessage
+) bool {
+	var foundMisbehaviour bool
+	switch msg.(type) {
+	case *Header:
+		foundMisbehaviour = false
+	case *Misbehaviour:
+		foundMisbehaviour = true
+	}
```
Comment (colin-axner) [inline modules/light-clients/06-solomachine/types/update.go:103]: ma
```suggestion
	switch msg.(type) {
	case *Misbehaviour:
		return true
	default: 
	    return false
	}
```
Will this work? And then you can remove the local var

Comment (damiannolan) [reply modules/light-clients/06-solomachine/types/update.go:103]: Yeah sure. I was also considering removing the type switch entirely and just using type assertion:

```
if _, ok := msg.(*Misbehaviour); ok {
    return true
}

return false
```

Either works really I guess!

File: modules/light-clients/06-solomachine/types/update.go [hunk 2]
```diff
@@ -84,6 +89,31 @@ func (cs ClientState) VerifyClientMessage(cdc codec.BinaryCodec, clientMsg expor
 	return nil
 }
 
+// CheckForMisbehaviour returns true for type Misbehaviour (passed VerifyClientMessage check), otherwise returns false
+func (cs ClientState) CheckForMisbehaviour(
```
[the remark this answers is not in the document]
Comment (damiannolan) [reply modules/light-clients/06-solomachine/types/update.go:93]: Great, thanks for confirming! :)

File: modules/light-clients/06-solomachine/types/update.go (hunk 1 above)
Comment (colin-axner) [reply modules/light-clients/06-solomachine/types/update.go:103]: That sounds good to me to!

**Comment by damiannolan** [comment]
> Needs basic tests

Added basic tests for `CheckForMisbehaviour` and put test function in place for `UpdateStateOnMisbehaviour` which will require more attention after storage updates 👍

File: modules/light-clients/06-solomachine/types/update.go [hunk 3]
```diff
@@ -84,6 +89,27 @@ func (cs ClientState) VerifyClientMessage(cdc codec.BinaryCodec, clientMsg expor
...
+	}
+
+	return false
+}
+
+// UpdateStateOnMisbehaviour updates state upon misbehaviour. This method should only be called on misbehaviour
+// as it does not perform any misbehaviour checks.
+func (cs ClientState) UpdateStateOnMisbehaviour(
+	_ sdk.Context, _ codec.BinaryCodec, _ sdk.KVStore, // prematurely include args for self storage of consensus state
+) (*ClientState, exported.ConsensusState, error) {
```
Comment (colin-axner) [inline modules/light-clients/06-solomachine/types/update.go:108]: I'm starting to wonder if we should not return an error? I feel like the code should panic as any errors detected should be returned earlier

Comment (damiannolan) [reply modules/light-clients/06-solomachine/types/update.go:108]: I agree. I was expecting `VerifyClientMessage` to return an error, `CheckForMisbehaviour` to return a bool and both `UpdateState` and `UpdateStateOnMisbehaviour` to be void.

Comment (seantking) [reply modules/light-clients/06-solomachine/types/update.go:108]: If we just return a bool from `CheckForMisbehavior` we miss out on the ability to return an error message specific to the misbehavior.

Comment (colin-axner) [reply modules/light-clients/06-solomachine/types/update.go:108]: > we miss out on the ability to return an error message specific to the misbehavior

Can you give an instance of this?

**Review by colin-axner — APPROVED** [review]
We can adjust the API of `UpdateStateOnMisbehaviour` and `CheckForMisbehaviour` later after we settle discussion

**Pull request merged**